File: backend/app/websocket/handler.py
from __future__ import annotations

from fastapi import APIRouter, WebSocket, WebSocketDisconnect

from ..models.messages import PongMessage, StatusMessage, StatusPayload
from ..pty.manager import pty_manager
from ..pty.sanitize import strip_terminal_identity_responses
from ..utils.keymapper import map_special_key
from .manager import ws_manager

router = APIRouter()


@router.websocket("/ws/terminal")
async def terminal_ws(websocket: WebSocket) -> None:
    await ws_manager.connect(websocket)

    try:
        for chunk in pty_manager.history_snapshot():
            cleaned = strip_terminal_identity_responses(chunk)
            if cleaned:
                await ws_manager.send_json(websocket, {"type": "output", "payload": {"data": cleaned}})

        st = pty_manager.status()
        await ws_manager.send_json(
            websocket,
            StatusMessage(
                type="status",
                payload=StatusPayload(status=st["status"], pid=st.get("pid"), message=None),
            ).model_dump(),
        )

        while True:
            data = await websocket.receive_json()
            msg_type = data.get("type")
            payload = data.get("payload") or {}

            if msg_type == "input":
                raw = str(payload.get("data", ""))
                cleaned = strip_terminal_identity_responses(raw)
                if cleaned:
                    pty_manager.write(cleaned)
                continue

            if msg_type == "resize":
                cols = int(payload.get("cols", 0) or 0)
                rows = int(payload.get("rows", 0) or 0)
                if cols > 0 and rows > 0:
                    pty_manager.resize(cols, rows)
                continue

            if msg_type == "special_key":
                seq = map_special_key(str(payload.get("key", "")), list(payload.get("modifiers", []) or []))
                if seq:
                    pty_manager.write(seq)
                continue

            if msg_type == "ping":
                await ws_manager.send_json(websocket, PongMessage(type="pong", payload={}).model_dump())
                continue

    except WebSocketDisconnect:
        pass
    finally:
        await ws_manager.disconnect(websocket)

File: backend-node/src/websocket/handler.ts
import WebSocket, { WebSocketServer } from 'ws';
import { Server as HttpServer } from 'http';
import { wsManager } from './manager.js';
import { ptyManager } from '../pty/manager.js';
import { stripTerminalIdentityResponses } from '../pty/sanitize.js';
import { mapSpecialKey } from '../utils/keymapper.js';
import { StatusMessage, PongMessage } from '../types/messages.types.js';

export function setupWebSocket(server: HttpServer): void {
  const wss = new WebSocketServer({ server, path: '/ws/terminal' });

  wss.on('connection', (ws: WebSocket) => {
    wsManager.connect(ws);

    // Send history replay
    for (const chunk of ptyManager.historySnapshot()) {
      const cleaned = stripTerminalIdentityResponses(chunk);
      if (cleaned) {
        wsManager.sendJson(ws, { type: 'output', payload: { data: cleaned } });
      }
    }

    // Send current status
    const st = ptyManager.status();
    const statusMsg: StatusMessage = {
      type: 'status',
      payload: {
        status: st.status,
        pid: st.pid,
        message: null,
      },
    };
    wsManager.sendJson(ws, statusMsg);

    // Handle incoming messages
    ws.on('message', (rawData) => {
      try {
        const data = JSON.parse(rawData.toString());
        const msgType = data.type;
        const payload = data.payload || {};

        if (msgType === 'input') {
          const raw = String(payload.data || '');
          const cleaned = stripTerminalIdentityResponses(raw);
          if (cleaned) {
            ptyManager.write(cleaned);
          }
        } else if (msgType === 'resize') {
          const cols = parseInt(payload.cols, 10) || 0;
          const rows = parseInt(payload.rows, 10) || 0;
          if (cols > 0 && rows > 0) {
            ptyManager.resize(cols, rows);
          }
        } else if (msgType === 'special_key') {
          const seq = mapSpecialKey(
            String(payload.key || ''),
            payload.modifiers || []
          );
          if (seq) {
            ptyManager.write(seq);
          }
        } else if (msgType === 'ping') {
          const pongMsg: PongMessage = { type: 'pong', payload: {} };
          wsManager.sendJson(ws, pongMsg);
        }
      } catch {
        // Ignore malformed messages
      }
    });

    ws.on('close', () => {
      wsManager.disconnect(ws);
    });

    ws.on('error', () => {
      wsManager.disconnect(ws);
    });
  });
}
